### ISSUE TITLE
JsonTextReader should be less greedy when reading from TextReader

### ISSUE THREAD

**Issue by AArnott** [JamesNK/Newtonsoft.Json#803]
The `JsonTextReader` seems to read large blocks of data from the `TextReader` fed into it, without regard to how much data is actually pulled through it. This is leading to difficulty in using it (via the `JsonSerializer`) in situations where the `TextReader`'s source is a stream (network, or named pipe).

There are two primary difficulties the current greedy reading presents:

1. If a stream contained both JSON and other content after it, someone reading the stream can't use the `JsonTextReader` to read the JSON and then read other data in the stream that follows the JSON because those bytes have already been read into the `JsonTextReader`'s buffer.
1. Because there are no async methods on the `JsonTextReader` and `JsonSerializer` (#650), reading objects from a network stream can block the calling thread for very long periods of time. This is unacceptable in some applications such as the one I'm working on. We tried to workaround this by asynchronously reading from the network stream first and copying into a `MemoryStream` which we then feed to the `JsonTextReader`, but at that level we don't know when we reach the end of a JSON object so we may have to go back and read more from the network stream. While it can be a very long time between messages such that an async wait is important, if a message is incomplete the rest can be expected to arrive soon after (in my app anyway) so even if #650 isn't done, if we could at least be sure the reader isn't greedy, we'd be able to add our own async buffer to allow us to asynchronously wait for messages and then (when necessary) synchronously block for the rest of the message to come in.

The below unit test fails demonstrates the greedy nature and would pass if `JsonTextReader` stopped reading from the `TextReader` once it encountered the closing object token. Ideally, it would read exactly the number of characters required for every kind of token. But at least for my application we'd benefit from it ensuring that it reads no further than the balanced EndObject token.

```csharp
using System;
using System.IO;
using System.Text;
using Newtonsoft.Json;
using Xunit;
using Xunit.Abstractions;

public class Parsers
{
    private const string json = "{'a':3}{'a':5}";
    private static readonly int[] objectFinishPositions = new[] { 7, 14 };
    private readonly ITestOutputHelper logger;
    private readonly Stream ms;
    private readonly StreamReader sr;

    public Parsers(ITestOutputHelper logger)
    {
        this.logger = logger;

        this.ms = new MemoryStream(Encoding.UTF8.GetBytes(json));
        this.sr = new StreamReader(ms);
    }

    [Fact]
    public void Newtonsoft()
    {
        var jr = new JsonTextReader(this.sr) { SupportMultipleContent = true };
        int objectCount = 0;
        while (jr.Read())
        {
            if (jr.TokenType == JsonToken.EndObject)
            {
                Assert.Equal(objectFinishPositions[objectCount], this.ms.Position); // this currently fails
                objectCount++;
            }
        }
    }
}
```

**Comment by JamesNK** [comment]
Adding async from the reader/writer level up to the serializer level would be an enormous amount of work and not something I want to think about right now.

If the problem is JsonTextReader is too greedy, how much less greedy should it be? I think right now it will full up its buffer. The buffer initial length is 1024 but can grow larger if required (reading a large string). Would a fix be to limit the amount of read data to a maximum limit rather than blindly filling the buffer? If so do you have a suggestion of what the max should be?

**Comment by AArnott** [comment]
Reading an arbitrary number of characters is problematic, no matter what that number of characters is. The key is reading one character at a time (including Peek() if necessary) in order to read exactly as many characters as necessary to complete a token. No more, no less.

**Comment by AArnott** [comment]
CC: @ilyabiryukov

**Comment by JamesNK** [comment]
I don't want to do that. Making a virtual method call to get a character at a time from the TextReader would hurt performance.

Maybe you should get a copy of the JsonTextReader.cs and modify it?

**Comment by AArnott** [comment]
Thanks for considering.
If I had a source copy of JsonTextReader, I wouldn't be able to pass an instance of my class to your `JsonSerializer` as the type would not have the same assembly identity.

Does JsonTextReader have virtual methods sufficient for me to derive from yours to customize the behavior?

**Comment by AArnott** [comment]
I'm guessing the answer to my question is no. So I guess we can consider this issue *Won't Fix* if you're pretty confident the perf impact would be significant and negative.

**Comment by JamesNK** [comment]
> If I had a source copy of JsonTextReader, I wouldn't be able to pass an instance of my class to your JsonSerializer as the type would not have the same assembly identity.

It's just a class that inherits from JsonReader. Copy and paste it, make changes, rename it to ArnottJsonReader and go crazy.

**Comment by AArnott** [comment]
Actually @ilyabiryukov suggested (offline) that because `JsonSerializer` actually accepts a `JsonReader` your suggestion should work after all. :)

**Comment by AArnott** [comment]
(messages crossed. :smile:) thanks again.

**Comment by AArnott** [comment]
Do you have any perf tests that we could use to compare ours to yours when we were done? (just out of curiosity)

**Comment by JamesNK** [comment]
PerformaceTests.cs

It like all the tests are messy.

**Comment by forteco** [comment]
Is there any way, that the Json Serializer will not throw exceptions when it Reads an Incomplete JSON?
I modified the sample above for illustration.
What I am doing is this: I am parsing a network-stream of Json objects that I receive. So from time
to time I get partially transmitted json objects. So if this happens, I want to parse for data when the
next package arrives, and then I have to be sure that the stream position is unchanged, so that
then it would be able to read a complete json object. Any ideas?
Thanks! 

```
public class JsonReaderTest
    {
        public class Role
        {
            public string Name { get; set; }
        }
        
        public static void Run()
        {
            string json = @"{ 'name': 'Admin' }{ 'name': 'Publisher' } {'name':";

            IList<Role> roles = new List<Role>();

            var stringReader = new StringReader(json);
            
            JsonTextReader reader = new JsonTextReader(stringReader);
            reader.SupportMultipleContent = true;
            reader.CloseInput = false;
            
            JsonSerializer serializer = new JsonSerializer();
           
            
            while (true)
            {
                if (!reader.Read())
                {
                    break;
                }

                try
                {
                    Role role = serializer.Deserialize<Role>(reader);
                    roles.Add(role);
                }
                catch (Exception ex)
                {
                    int j = 55;
                }
            }

            foreach (Role role in roles)
            {
                Console.WriteLine(role.Name);
            }
        }
```

**Comment by forteco** [comment]
Thanks James! Unfortunately I didnt manage to understand your code (yet).
Basically if json deserialize fails (throws exception) then one has to move
backwards in the stream for the size of the uncomplete fragment and then
read again when the next networkstream read unblocks.  I think this is
exactly what you do.  In dont think it has to do anything with async
processing though.

**Comment by forteco** [comment]
Sorry!  Thanks ANDREW!

**Comment by AArnott** [comment]
@forteco My approach still requires that you know when you have a complete JSON fragment that can be deserialized. I do this in any of several ways including prepending the length before the JSON, or (if the serializer don't know the length before writing to the network) [using substreams](https://github.com/AArnott/Nerdbank.Streams/blob/master/doc/Substream.md). That way I don't have to keep trying to deserialize until it happens to not throw an exception.

**Comment by forteco** [comment]
@AArnott  It seems we are working on very similar problems :-) I have implemented a data-feed that uses StreamingJSON over http. And I am basically receiving always the same data from that stream.
I basically read the entire available data as string, and then Json-Convert it, and then I use a leftover
string that I insert when the next package arrives. I manage to detect partial json objects by catching Json Serializer events. My approach works only for string-parsing that does not contain \r \n because
I use the LinePosition to determine how much of the data was read. I tried using reflection to get
more sensible read-position but I failed. I think the other approach might be to parse JToken, which
I think gets information if the json was complete or not, and then somehow pass that JToken to the serializer. But I havent been able to get that to work so far.

```
 public static int Parse(MLogger logger, string sData, Data dataCallback,  bool Debug)
        {
            MemoryStream ms = new MemoryStream();
            StreamWriter sw = new StreamWriter(ms);
            sw.Write(sData);
            sw.Flush();
            ms.Flush();
            ms.Position = 0;
            using (var sr = new StreamReader(ms))
            {
                try
                {
                    var serializer = new Newtonsoft.Json.JsonSerializer();
                    serializer.CheckAdditionalContent = false;
                    using (var jsonReader = new Newtonsoft.Json.JsonTextReader(sr))
                    {
                        jsonReader.SupportMultipleContent = true;
                        jsonReader.CloseInput = false;
                        return ReadJsonFromReader(logger, sData, serializer, jsonReader, dataCallback, Debug);
                    }
                }
                catch (Exception ex)
                {
                    logger.Info("JsonParser exception: " + ex.Message);
                    return 0;
                }
            }
        }


        static int ReadJsonFromReader(MLogger logger, string sData, Newtonsoft.Json.JsonSerializer serializer, Newtonsoft.Json.JsonTextReader jsonReader, Data dataCallback, bool Debug)
        {
            int readUntil = 0;
            
            TcpFeedData data; 
           
            while (jsonReader.Read())  
            {
                data = null;
                bool hadEx = false;
                try
                {
                    //logger.Info($"JsonReader TokenType: {jsonReader.TokenType} "); // this is always StartObject; no use to use it.
                    data = serializer.Deserialize<TcpFeedData>(jsonReader);
                }
                catch (Newtonsoft.Json.JsonReaderException jrex)
                {
                    data = null;
                    hadEx = true; // happens when json is not complete.
                    if (Debug) logger.Info("JsonParser occured leftover.");
                }
                catch (Exception ex)
                {
                    data = null;
                    hadEx = true; // happens when json is not complete.
                    if (Debug) logger.Info("JsonParser occured leftover/ other " + ex.Message);
                }
                
                // SEND BACK DATA EVENTS
                try
                {
                    if (data != null)
                    {
                        string jsonReceived = null;
                        if (data.IsLive())
                        {
                            // If Multiple JsonObjects are parsed, then the json will contain all the elements 
                            int jsonStart = readUntil;
                            int jsonEnd = jsonReader.LinePosition; // CharsUsed(jsonReader);
                            if (jsonEnd > jsonStart)
                            {
                                jsonReceived = sData.Substring(jsonStart, jsonEnd - jsonStart);
                            }
                        }

                        dataCallback(data, jsonReceived);
                    }
                }
                catch (Exception ex)
                {
                    logger.Info("JsonParser dataCallback exception: " + ex.Message + " stack: " + ex.StackTrace);
                }
                
                //logger.Info($"JsonReader strLen: {sData.Length} LinePosition: {jsonReader.LinePosition}");
                if (!hadEx)
                {
                    readUntil = jsonReader.LinePosition;
                }
                
            }
            //readUntil= jsonReader.LinePosition;
            //readUntil = CharsUsed(jsonReader);
            if (Debug)
            {
                logger.Info($"JsonParser used {readUntil} of {sData.Length}.");
            }
            
            return readUntil;
        }
```

**Comment by AArnott** [comment]
Well, if you can't control the other side of the stream so that it emits length headers of some sort, then perhaps what you're doing is the best option available.
